### PR TITLE
*: support metric.

### PR DESCRIPTION
**Pull request by overvenus**
Issue: #566 

Plan b: 
 - A new field `metric` in `struct Server`.
 - `metric` wraps `cadence`, provides a statsd client.

### CODE REVIEW

**Comment by ngaut** [comment]
Cool @overvenus 
How does it compare to the unsafe version that you mentioned before?

File: src/util/metric.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+impl Metric {
+    pub fn new(prefix: &str, level: LogLevel) -> Metric {
+        let sink = LoggingMetricSink::new(level);
+        Metric { inner: StatsdClient::from_sink(prefix, sink) }
+    }
+
+    // Increment the counter by `1`
+    pub fn incr(&self, key: &str) {
+        let r = self.inner.incr(key);
```
Comment (siddontang) [inline src/util/metric.rs]: use `if let Err(e) = ` if you only need to handle error.

File: src/server/server.rs [hunk 2]
```diff
@@ -77,6 +78,10 @@ pub struct Server<T: RaftStoreRouter + 'static, S: StoreAddrResolver> {
     end_point: EndPointHost,
 
     resolver: S,
+
+    // TODO: remove `#[allow(dead_code)]`
+    #[allow(dead_code)]
+    metric: Metric,
```
Comment (siddontang) [inline src/server/server.rs]: I think you should use Arc<Metric>, we may use it in other places, like rafestore.

**Comment by overvenus** [comment]
I want use a static global variable `StatsdClient` like LOGGER in **crate log** at the first try, so that APIs like `count()`, `time()` can be called in any function, but after a mountain of search, I found that `unsafe` is inevitable.

In this PR `StatsdClient` lives in `Server` which will live as long as the program runs, but APIs may be called with assistance of `Server`, eg:
```rust
fn foo(&self) { // self: Server
    self.metric.count("bar", 42);
}
```
or `svr.metric.count("bar", 42);`

**Comment by ngaut** [comment]
PTAL @BusyJay

File: src/util/metric.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use log::LogLevel;
+use cadence::prelude::{Counted, Timed, Gauged, Metered};
```
Comment (BusyJay) [inline src/util/metric.rs]: Why do we need these?

File: src/util/metric.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+use cadence::{StatsdClient, LoggingMetricSink};
+
+pub struct Metric {
```
Comment (BusyJay) [inline src/util/metric.rs]: Why do we need to wrap the client?

File: src/util/metric.rs (hunk 3 above)
Comment (overvenus) [inline src/util/metric.rs]: It wont compile without these.
`Counted, Timed, Gauged, Metered` are triats that contain some method signatures used below.

eg: `Time` => https://github.com/pingcap/tikv/pull/577/files#diff-fa8093341d2e5ce6a6fb6bb4fbb1affeR50

File: src/util/metric.rs (hunk 4 above)
Comment (overvenus) [inline src/util/metric.rs]: It will be convenient to change the underlying metric implemention in the future.

File: src/util/metric.rs (hunk 3 above)
Comment (BusyJay) [inline src/util/metric.rs]: Then import from cadence directly. `prelude` is often used with `*`.

File: src/util/metric.rs (hunk 4 above)
Comment (BusyJay) [inline src/util/metric.rs]: If we share the same method signature with `StatsdClient`, I don't see any point doing it right now. As we may not change the implementation at all in the end.

Comment (overvenus) [inline src/util/metric.rs]: So I only need to add a new field `Arc<StatsdClient<LoggingMetricSink>>` in `Server`?

If so, I will delete the metric.rs, and use cadence directly.

**Comment by ngaut** [comment]
@BusyJay Do you prefer using static methods ?

File: src/bin/tikv-server.rs [hunk 5]
```diff
@@ -206,6 +217,14 @@ fn main() {
                 "log",
                 "set log level",
                 "log level: trace, debug, info, warn, error, off");
+    opts.optopt("M",
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: We may need a way to be configured using UDP sink.

**Comment by BusyJay** [comment]
I think both are ok. But I prefer using macro, it's clean and neat.

**Comment by overvenus** [comment]
@BusyJay @ngaut 
As far as I know, this feature may require a static variable, like LOGGER in `crate log`, that will introduce `unsafe` if we choose macro or static methods. 
As I'm newbie to rust I'd like to hear comments from you if you know some crate use static variables w/o `unsafe`.

BTW, I can come up with some marco or static methods that create statsd client on the fly, but it is definitely unacceptable if we use udp.

**Comment by overvenus** [comment]
See #595

**Pull request closed without merging**